File: tests/typesafety/typescript/unit_arithmetic_misuse.ts
// expect: fails

declare const metersBrand: unique symbol;
declare const kilogramsBrand: unique symbol;

type Meters = number & { readonly [metersBrand]: true };
type Kilograms = number & { readonly [kilogramsBrand]: true };

function meters(n: number): Meters {
  return n as Meters;
}

function kilograms(n: number): Kilograms {
  return n as Kilograms;
}

function addMeters(a: Meters, b: Meters): Meters {
  return (a + b) as Meters;
}

export const nonsense = addMeters(meters(1.8), kilograms(4)); // error: Kilograms is not Meters

File: tests/typesafety/python/unit_arithmetic_misuse.py
# expect: fails
# title: Seconds plus Kilograms, rejected by the checker
# misuse-of: unit_arithmetic
from dataclasses import dataclass


@dataclass(frozen=True)
class Seconds:
    value: float

    def __add__(self, other: "Seconds") -> "Seconds":
        return Seconds(self.value + other.value)


@dataclass(frozen=True)
class Kilograms:
    value: float

    def __add__(self, other: "Kilograms") -> "Kilograms":
        return Kilograms(self.value + other.value)


def nonsense(duration: Seconds, load: Kilograms) -> Seconds:
    return duration + load  # error: Seconds + Kilograms has no meaning
